### PR TITLE
strongswan: migrate to swanctl configs

### PR DESCRIPTION
**Pull request by pprindeville**
Maintainer: me, @Thermi
Compile tested: x86_64, generic, head (7d12f29ae1)
Run tested: same, running on a production router

Description:

Derived from the ipsec initd script, with the following changes:

(1) various code improvements, corrections (get rid of left/right updown scripts, since there's only one), etc;
(2) add reauth and fragmentation parameters;
(3) add x.509 certificate-based authentication;
(4) add xfrm dependency;
(5) drop force_crypto_proposal option;
(6) make the include in swanctl.conf be persistent;

and other minor changes.

Requires PR #14711.

### CODE REVIEW

**Comment by stintel** [comment]
Please drop the xfrm dependency. You already have a PR open for this, and I told you I'm inclined to NACK it. It's also completely unrelated to the new init script.

You also did not include the new init script in any package, so this change is pretty pointless.

**Comment by pprindeville** [comment]
> You also did not include the new init script in any package, so this change is pretty pointless.

@stintel Sorry, got dropped during a rebase.  Added it back.

File: net/strongswan/Makefile
Comment (stintel) [inline net/strongswan/Makefile:533]: Missing an $(INSTALL_DIR) $(1)/etc/init.d. Preferably add it at the top of the INSTALL macros of this package, so that they are somewhat in alphabetical order.

**Review by stintel — DISMISSED** [review]

**Comment by stintel** [comment]
I'll need some time to properly test this, for now I would like to ask to not merge this without my approval.

File: net/strongswan/Makefile
Comment (pprindeville) [reply net/strongswan/Makefile:534]: Done.

**Comment by feckert** [comment]
@pprindeville Great work.
I am also in the process of setting up a setup and testing your pullrequest for swanctl migration.
Just two notes in advance about the codeing style.
* So I do not always have to count the blanks when creating the configuration can not we do that in the `swanctl_xappend`?  Take a look at this in [keepalived](https://github.com/openwrt/packages/blob/master/net/keepalived/files/keepalived.init) as a template. This would make this code area clearer https://github.com/openwrt/packages/pull/14708/commits/5cd7672d67d381b2d6a89455281e2293553c09b2#diff-54d296a56c8c4d6e287673b3f5aa2c037abad2a8ae91f37314ab435d1dc4db28R205
* Also could we remove all this white spaces between `$1` in https://github.com/openwrt/packages/pull/14708/commits/5cd7672d67d381b2d6a89455281e2293553c09b2#diff-54d296a56c8c4d6e287673b3f5aa2c037abad2a8ae91f37314ab435d1dc4db28R135 This is a matter of taste. But I find it better to read.

That important, whether it all works I still have to test

**Comment by pprindeville** [comment]
> @pprindeville Great work.
> I am also in the process of setting up a setup and testing your pullrequest for swanctl migration.
> Just two notes in advance about the codeing style.
> 
> * So I do not always have to count the blanks when creating the configuration can not we do that in the `swanctl_xappend`?  Take a look at this in [keepalived](https://github.com/openwrt/packages/blob/master/net/keepalived/files/keepalived.init) as a template. This would make this code area clearer [5cd7672#diff-54d296a56c8c4d6e287673b3f5aa2c037abad2a8ae91f37314ab435d1dc4db28R205](https://github.com/openwrt/packages/commit/5cd7672d67d381b2d6a89455281e2293553c09b2#diff-54d296a56c8c4d6e287673b3f5aa2c037abad2a8ae91f37314ab435d1dc4db28R205)
> * Also could we remove all this white spaces between `$1` in [5cd7672#diff-54d296a56c8c4d6e287673b3f5aa2c037abad2a8ae91f37314ab435d1dc4db28R135](https://github.com/openwrt/packages/commit/5cd7672d67d381b2d6a89455281e2293553c09b2#diff-54d296a56c8c4d6e287673b3f5aa2c037abad2a8ae91f37314ab435d1dc4db28R135) This is a matter of taste. But I find it better to read.
> 
> That important, whether it all works I still have to test

Done and done.

File: net/strongswan/Makefile
Comment (pprindeville) [reply net/strongswan/Makefile:534]: > Missing an $(INSTALL_DIR) $(1)/etc/init.d. Preferably add it at the top of the INSTALL macros of this package, so that they are somewhat in alphabetical order.

@stintel Please confirm that this has been done to your satisfaction.

**Comment by pprindeville** [comment]
@neheb And have a look at this too, please.

**Comment by pprindeville** [comment]
> I'll need some time to properly test this, for now I would like to ask to not merge this without my approval.

@stintel Can you please acknowledge that your requested change has been done?

**Comment by stintel** [comment]
> > I'll need some time to properly test this, for now I would like to ask to not merge this without my approval.
> 
> @stintel Can you please acknowledge that your requested change has been done?

I can, but I still would like to test this. When I rebooted my APU2, my /etc/ipsec.conf and /etc/ipsec.secrets where gone, and only contained the include line generated by the init script. Because of this, my VPN connection was broken. Fortunately this happened on my local router. But this means something is broken and this needs to be solved first, before I'm going to accept changes that potentially break things even further. I'm on holiday until 6/03.

Please consider this as an explicit NACK until I am able to test. If somebody merges things without my approval and things break, I will seek to have commit access removed. I warned many times before, breaking existing setups is not acceptable, especially in times with extreme travel restrictions, as remote setups might become unreachable with no ability to go on-site to fix things.

**Comment by pprindeville** [comment]
> I can, but I still would like to test this. When I rebooted my APU2, my /etc/ipsec.conf and /etc/ipsec.secrets where gone, and only contained the include line generated by the init script. Because of this, my VPN connection was broken. Fortunately this happened on my local router. But this means something is broken and this needs to be solved first, before I'm going to accept changes that potentially break things even further. I'm on holiday until 6/03.

Was this a freshly imaged machine, or one that had a package update?

**Comment by pprindeville** [comment]
@stintel Better yet, send me your public key.  I have a VM (x86_64, BIOS, 2 Ethernet interfaces) that we can use to see if we can reproduce this issue...  but I suspect it will go away if we merge PR #14711 in first.

**Comment by stintel** [comment]
> @stintel Was this a freshly imaged machine, or one that had a package update?

I use only self-built images with sysupgrade, no package updates. I have not rebooted since. I hope to spend some time on further debugging/testing/reviewing stuff this weekend.

**Comment by pprindeville** [comment]
PR #14711 which is a dependency has been merged, so we're now unblocked.

File: net/strongswan/Makefile
Comment (pprindeville) [reply net/strongswan/Makefile:547]: @stintel If all the changes you've requested have been made, please approve.

**Comment by pprindeville** [comment]
@Thermi please review.

File: net/strongswan/files/swanctl.init
Comment (Thermi) [inline net/strongswan/files/swanctl.init:143]: Please reorder so it's encryption-hash-prf-dh_group
And if encryption is an AEAD algorithm, reject proposals with a hash, because that'd be an invalid proposal in IKE.

File: net/strongswan/files/swanctl.init
Comment (Thermi) [inline net/strongswan/files/swanctl.init:534]: We shouldn't generate the default file contents again, just the ones that are not available to the user to edit at any time (so effectively just the settings we want).
That way the user can decide if he/she wants to include the default config files, or not.
Effectively that means removing load_modular and the the plugins include line from here.

File: net/strongswan/files/swanctl.init
Comment (pprindeville) [reply net/strongswan/files/swanctl.init]: I've added the dh group support back.  The other changes of inheriting the IKE DH groups into ESP are more involved.  I suggest adding this as a subsequent PR.

File: net/strongswan/files/swanctl.init
Comment (pprindeville) [reply net/strongswan/files/swanctl.init:418]: I'm not sure v1 is worth supporting. Who doesn't support v2 at this point?

File: net/strongswan/files/swanctl.init
Comment (pprindeville) [reply net/strongswan/files/swanctl.init:536]: We're generating the entire profile via UCI, so the user doesn't have to hand-edit anything.

Comment (Thermi) [reply net/strongswan/files/swanctl.init:536]: I am talking about that the file we're creating is `/var/ipsec/strongswan.conf` and we're adding `"\ninclude /var/ipsec/strongswan.conf"` to `/etc/strongswan.conf`.
`/etc/strongswan.conf` default content already contains `charon.load_modular=yes` and the include line for `/etc/strongswan.d/*.conf`.
`/etc/strongswan.d/charon.conf` effectively contains the following line, too, already:
```
charon {
plugins {
include /etc/strongswan.d/charon/*.conf
}
```

If the user doesn't want that, he/she removed that. And then we add that back in via `/var/ipsec/strongswan.conf`. I think the user knows better in this regard.
We should just remove these includes from our generated file, especially given that we don't use any of the files that are included by the include directives in the generated file.

File: net/strongswan/files/swanctl.init
Comment (Thermi) [reply net/strongswan/files/swanctl.init:418]: *points at majority of vendors*
Also, the majority don't support that for IKEv2 either.

File: net/strongswan/files/swanctl.init
Comment (Thermi) [reply net/strongswan/files/swanctl.init]: As long as it's in the next release of openwrt or the next package version, that's fine.

File: net/strongswan/files/swanctl.init
Comment (pprindeville) [reply net/strongswan/files/swanctl.init:248]: Okay, what sort of warning?

File: net/strongswan/files/swanctl.init
Comment (pprindeville) [reply net/strongswan/files/swanctl.init:536]: Okay, removed:

```
  load_modular = yes
  ...
  plugins {
    include /etc/strongswan.d/charon/*.conf
  }
```

File: net/strongswan/files/swanctl.init
Comment (Thermi) [inline net/strongswan/files/swanctl.init:542]: Please replace `/usr/sbin/swanctl --load-all` with `/usr/sbin/swanctl --load-all --noprompt` to make sure `swanctl` does not try to prompt for passphrases for encrypted keys.

File: net/strongswan/files/swanctl.init
Comment (Thermi) [inline net/strongswan/files/swanctl.init:568]: Same, please add --noprompt.

File: net/strongswan/files/swanctl.init
Comment (pprindeville) [reply net/strongswan/files/swanctl.init:418]: I'm okay with making IKE v2 full-featured, but let's add onto it via an enhancement PR.  I want to get this out into the hands of users soon.

File: net/strongswan/files/swanctl.init
Comment (Thermi) [reply net/strongswan/files/swanctl.init:248]: E.g. if `closeaction` is not `clear`,  `WARNING: closeaction=$closeaction can cause instability`

File: net/strongswan/files/swanctl.init
Comment (pprindeville) [reply net/strongswan/files/swanctl.init:402]: Uh... not sure why it got duplicated in `config_conn()` and `config_remote()`.  Removing it from the former.

File: net/strongswan/files/swanctl.init
Comment (pprindeville) [reply net/strongswan/files/swanctl.init:248]: Added.

File: net/strongswan/files/swanctl.init
Comment (pprindeville) [reply net/strongswan/files/swanctl.init]: Done.

**Review by Thermi — APPROVED** [review]

**Pull request merged**
